### ISSUE TITLE
Docker Registry SSL CA cert location on windows?

### ISSUE THREAD

**Issue by justechn** [docker/docker#21189]
I am trying to follow this guide https://docs.docker.com/engine/security/certificates/, but the paths are for linux. What would be the equivalent for Windows?

**Comment by justechn** [comment]
@GordonTheTurtle this is not a bug report, I am simply trying to get a little help using this product.

**Comment by justechn** [comment]
Any help here would be appreciated.

**Comment by friism** [comment]
Here's the code for Linux versus Windows:

 * https://github.com/docker/docker/blob/12c67f42d85103c57c14f27579ff46fccdc3ee07/registry/config_windows.go#L10
 * https://github.com/docker/docker/blob/12c67f42d85103c57c14f27579ff46fccdc3ee07/registry/config_unix.go

So it looks like it's `os.Getenv("programdata") + `\docker\certs.d``. I haven't tested this though.

If you experiment with this and get this working, it'd be awesome if you you'd be interested in contributing to the doc: https://github.com/docker/docker/blob/eb729a0cb0be1f2d2ba5c616fcd4e585289f6dfc/docs/security/certificates.md

**Comment by justechn** [comment]
This would probably work, except that according to the docs we also have to include the hostname in the directory tree. As you can see here https://docs.docker.com/registry/deploying/ you are recommending that we use port 5000, which is included in the hostname (localhost:5000). So my full directory tree would be C:\ProgramData\docker\certs.d\localhost:5000\ca.crt

However Windows does not support directory names with a colon in them. So I am not able to get the cert to load.

Once I get this working I will contribute the solution to the docs.

**Comment by justechn** [comment]
nevermind I see on the next line down https://github.com/docker/docker/blob/12c67f42d85103c57c14f27579ff46fccdc3ee07/registry/config_windows.go#L10 you do a replace on : I will try that and report back

**Comment by friism** [comment]
hm, strictly speaking, the port number is not part of the hostname, but maybe the docs are unclear here: https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax

**Comment by justechn** [comment]
I tried both C:\ProgramData\docker\certs.d\docker-registry.lan and C:\ProgramData\docker\certs.d\docker-registry.lan5000 and neither is working. I will keep playing around.

BTW I was able to get it working by using docker-machine ssh default. Creating the directory /etc/docker/certs.d/docker-registry.lan:5000 and creating the cert with vi ca.crt and pasting the cert in the file.

**Comment by justechn** [comment]
It would be really great to see what was being returned on this line.  https://github.com/docker/docker/blob/1d094255d57090d34b2c11c12b253c277ae1ce28/registry/registry.go#L35

I am guessing that it is not what we think it is.

is there any way to enable the debug that is specified here?
https://github.com/docker/docker/blob/1d094255d57090d34b2c11c12b253c277ae1ce28/registry/registry.go#L36

**Comment by friism** [comment]
@justechn does `-D` (when starting the engine) work?

**Comment by justechn** [comment]
yes that spit out a bunch of debug data, but nothing about "hostDir". I tried running a few commands, but those don't print out any debug.

**Comment by friism** [comment]
@justechn sorry I didn't follow up - I wanted to check in to see if you got this working?

**Comment by justechn** [comment]
No, I never got this working. For now I am loading the cert directly on the docker-machine.

**Comment by StefanScherer** [comment]
@justechn There is a PowerShell script to generate TLS certs and deploy them to the right place on a TP4 machine. Have a look at https://github.com/Microsoft/Virtualization-Documentation/tree/master/windows-server-container-tools/DockerTLS

The files `ca.pem`, `key.pem`, `server-key.pem` and `server-cert.pem` should be placed in the directory `c:\ProgramData\docker\certs.d\`. Then restart the docker service with `restart-service docker` from an admin powershell.

**Comment by StefanScherer** [comment]
The `C:\ProgramData\docker\runDockerDaemon.cmd` script shows the details how it switches from insecure to secure mode.

**Comment by StefanScherer** [comment]
Sorry, haven't seen that you try to use a secure Registry. Haven't done that yet.

**Comment by dls314** [comment]
I see this issue in Docker For Windows 1.12.0-rc2-beta16 when trying to use a secure registry on another host as a client. Has anyone gotten this case to work?

**Comment by Alex-Random** [comment]
yeah i think its broken on windows, it never loads the cert from that location

**Comment by StefanScherer** [comment]
I tried this with a Windows Docker daemon 1.12.0-rc4 on TP5 using steps in https://docs.docker.com/registry/insecure/#/using-self-signed-certificates and running the registry on Linux somewhere with https://docs.docker.com/registry/deploying/#get-a-certificate

After copying the `domain.crt` into `C:\ProgramData\docker\certs.d\myregistrydomain.com5000\ca.crt` without the colon it worked. So the path works for Windows Docker daemon.

I guess you want to use it with Docker for Windows running a Linux docker daemon?

**Comment by hofsa** [comment]
I'm trying to use a company-registry from "Docker for Windows 1.12.0-stabe Build 5968". The registry uses TLS and authentication with username/password.

When I'm trying to login from my "Docker for Windows"-CLI, I'm getting the following error-message:

docker login registry.mycompany.com:5000
Username: test
Password:
Error response from daemon: Get https://registry.mycompany.com:5000/v1/users/: x509: certificate signed by unknown authority

I copied my ca.cert to the following locations:
C:\Program Files\Docker\Docker\resources\certs.d\registry.mycompany.com5000
C:\Program Files\Docker\Docker\certs.d\registry.mycompany.com5000
C:\Program Files\Docker\certs.d\registry.mycompany.com5000

Looks like the docker daemon does not use the CA-Certificate from any of these locations.

@StefanScherer: Is this the same scenario you had?

**Comment by thaJeztah** [comment]
@hofsa did you restart the daemon after copying/adding the certificates?

**Comment by hofsa** [comment]
@thaJeztah Of course. I even rebooted the Windows-Machine, just to be sure.

Did you get this running? Can you post your registry-url and your directory-path?

**Comment by thaJeztah** [comment]
I'm not running on Windows, just thought I'd check if you restarted the daemon (otherwise it won't / may not load new certificates)

**Comment by phkrug** [comment]
I'm having the same problem with Docker for Windows and a self-signed certificate. Tried every suggested method on solving this issue without any success. docker-compose doesn't even seem to look into the C:\ProgramData\docker\cert.d path.

Is there any solution to this problem yet?

**Comment by hofsa** [comment]
I opened [Issue #36](https://github.com/docker/for-win/issues/36), where I also documented a work-around with the "insecure-registries"-option.

Hope that helps...

**Comment by felipeleon73** [comment]
I wrote a script based on that described by Docker Saigon [.](http://docker-saigon.github.io/post/Docker-Beta/)
In the same folder as the script you must create the certs.d folder with all its ramifications.

`$dockercerts="/host/etc/docker/certs.d"
$TEMPHOST="TEMP"

docker run --privileged --rm -v /var/run/docker.sock:/var/run/docker.sock -v /usr/bin/docker:/usr/bin/docker alpine docker run --net=host --ipc=host --uts=host --pid=host --security-opt=seccomp=unconfined --privileged -d --name $TEMPHOST -v /:/host alpine tail -f /dev/null
docker exec TEMP rm -r $dockercerts
docker exec TEMP mkdir $dockercerts
$directories = Get-ChildItem ".\certs.d"
foreach ($directory in $directories)
{
    $hostname=$directory.FullName
    $dockername=$dockercerts+"/"+$directory.Name.Replace('_',':')
    echo $hostname
    echo $directory.FullName.Replace('_',':')
    docker cp $hostname TEMP:$dockername
}
docker stop $TEMPHOST
docker rm $TEMPHOST`

**Comment by GHrw62** [comment]
Appreciate the information but it seems it is all for Windows 10.  I am using Windows 7 Enterprise and have the same issue "docker: Error while pulling image: Get https://index.docker.io/v1/repositories/library/hello-world/images: x509: certificate signed
by unknown authority."

tried server different solutions but none of the work.

**Comment by garyrusso** [comment]
I'm having the same problem. For Windows 7 Enterprise, I see that the pem files are in the User directory. C:\Users\janedoe\.docker\machine\certs

I'm going to see if copying the revised pem files to this directory will work.

**Comment by cs94njw** [comment]
Ditto same problem :( 
My company has an internal registry server, with their own generated certificate.
As far as I can tell, my configuration is good, but I also get:
`$ docker login hki-repo03.xxx.com:8222
Username: cpxxwr
Password:
Error response from daemon: Get https://hki-repo03.xxx.com:8222/v1/users/: x509: certificate signed by unknown authority`

**Comment by cs94njw** [comment]
If you're using Docker Toolbox, then THIS article solved it for me: https://github.com/boot2docker/boot2docker#insecure-registry
When essentially you install the certificates to your docker machine, instead of to your Windows installation.

Generate certs (I use the Docker Quickstart Terminal)
`openssl s_client -connect hki-repo03.xxx.com:8222`

Take bit of output from "BEGIN CERTI" to "END CERT" *INCLUSIVE**, and copy to “ca.crt”, e.g.

`-----BEGIN CERTIFICATE-----`
`MIIC6TCCAlICCQDjDRoI8gkx1TANBgkqhkiG9w0BAQUFADCBuDELMAkGA1UEBhMC`
`RkkxFjAUBgNVBAgMDUhlbHNpbmtpIEFyZWExETAPBgNVBAcMCEhlbHNpbmtpMRww`
`GgYDVQQKDBNDb21wdGVsIENvcnBvcmF0aW9uMRQwEgYDVQQLDAtJVCBTZXJ2aWNl`
`czEfMB0GA1UEAwwWaGtpLXJlcG8wMy5jb21wdGVsLmNvbTEpMCcGCSqGSIb3DQEJ`
`ARYaU3RveWFuLkFuZ2Vsb3ZAQ29tcHRlbC5DT00wHhcNMTYwMzEyMDk1MzAxWhcN`
`MTcwMzEyMDk1MzAxWjCBuDELMAkGA1UEBhMCRkkxFjAUBgNVBAgMDUhlbHNpbmtp`
`IEFyZWExETAPBgNVBAcMCEhlbHNpbmtpMRwwGgYDVQQKDBNDb21wdGVsIENvcnBv`
`cmF0aW9uMRQwEgYDVQQLDAtJVCBTZXJ2aWNlczEfMB0GA1UEAwwWaGtpLXJlcG8w`
`My5jb21wdGVsLmNvbTEpMCcGCSqGSIb3DQEJARYaU3RveWFuLkFuZ2Vsb3ZAQ29t`
`cHRlbC5DT00wgZ8wDQYJKoZIhvcNAQEBasdasdasQADgY0AMIGJAoGBAKdbrtYFjWugtG9u`
`VfqGkkR+i8q188MSfrFKE4Wzr0m7f3wasdlSsMx08Tv5C+U56z2zmlsojvBJFHY9k`
`D6b1V5WRtC2suysF5v/7bqsM1xdsadaRmavYU0rBhsaGaOLZ1d3Q3RkWIlKexShNc30R`
`LOXwf0FO0II4l1ltnNsq7LutasdasdAgMBAAEwDQYJKoZIhvcNAQEFBQADgYEAY/4V`
`udI6NItfUFdJoq3aiQ/qIsMiP0w1DutjgFe0l5bgObvt+Lo/lLQKxhKLy+whdhje`
`F2jETrm5Ti5bLLoFbPdGaaSFTboMmP/wjAyGOOtGv3xcKxc7zAa9SDgq4T6G1u5z`
`sMi+tXN0mv0HVpZ2CdSBKJ271FPSWWebYZnXY1E=`
`-----END CERTIFICATE-----`

Then run the following on Windows:

`docker-machine scp ca.crt default:ca.crt`
`docker-machine ssh default`
`sudo mkdir /etc/docker/certs.d/hki-repo03.xxx.com:8222/`
`sudo mv ~/ca.crt /etc/docker/certs.d/hki-repo03.xxx.com:8222/ca.crt`
`exit`
`docker-machine restart`

After that, the following said "login succeeded"

`docker login login hki-repo03.xxx.com:8222`

**Comment by fraccy** [comment]
I got round this on Windows 10 by running Certmgr.exe (Certificate Manager Tool), importing my self-signed certificate, then restarting docker (1.12-something). After this I was able to do docker login to my registry and lost the `x509: certificate signed by unknown authority` error.

**Comment by bevislu** [comment]
@fraccy Your solution solve my problem.

**Comment by MartinEden** [comment]
I can confirm that on my Windows 10 machine, adding certificates to C:\ProgramData\Docker\certs.d had no effect. In order to get it to work I had to use Windows global certificate store.

Here are some more detailed steps to expand on @fraccy's answer:
* Start > "Manage Computer Certificates" (also available in the control panel)
* Right-click on "Trusted Root Certification Authoritites" > "All tasks" > "Import"
* Browse to the crt file and then keep pressing "Next" to complete the wizard
* Restart Docker for Windows

**Comment by friism** [comment]
cc @jimgalasyn we should document this.

**Comment by MohsenBzm** [comment]
Just double click the `.crt` file and click next to the end.

**Comment by hishwest** [comment]
@MartinEden : which certficate you import please ?
thx in advance

**Comment by heiliuer** [comment]
@cs94njw It works from your idea !

**Comment by Elderry** [comment]
I use docker engine and above solutions seems don't work for me. On my macOS I have succeeded, then I 

- copied the files `client.cert`, `client.key` and `ca.crt` to my win 10 into:

    - `C:\ProgramData\Docker\certs.d\docker.company.net\`
    - `C:\Users\<user>\.docker\certs.d\docker.company.net\`

- imported the certificates into my Windows global certificates store and user certificates store.
Sadly, I still get this:
```
docker login docker.tradeshift.net
Username: user
Password:
Error response from daemon: Get https://docker.company.net/v2/: remote error: tls: handshake failure

docker search docker.company.com/hello-world
Error response from daemon: invalid registry endpoint https://docker.traeshift.com/v1/: Get https://docker.traeshift.com/v1/_ping: dial tcp: lookup docker.traeshift.com on 192.168.65.1:53: no such host. If this private registry supports only HTTP or HTTPS with an unknown CA certificate, please add `--insecure-registry docker.traeshift.com` to the daemon's arguments. In the case of HTTPS, if you have access to the registry's CA certificate, no need for the flag; simply place the CA certificate at /etc/docker/certs.d/docker.traeshift.com/ca.crt
```

**Comment by hfarukuslu** [comment]
in linux the path to put the client certificates is
/etc/docker/certs.d/<registry url>/

Question:
Where is this path in windows?

NOTE:  "C:\ProgramData\Docker\certs.d" **DOES NOT WORK**

**Comment by penleywang** [comment]
After testing, the path on the Windows is:
 %USERPROFILE%\.docker\certs.d
For Instance：
C:\Users\Penley.Wang\.docker\certs.d\reg.penley.wang\ca.crt
Then you will restart the Docker Desktop.

**Comment by minhhungit** [comment]
maybe you no need to find the location on windows, but can install cert on windows by this way:

Windows Server:

Open Windows Explorer, right-click the ***.crt** file, and choose Install certificate. 
When prompted, select the following options:

Store location: local machine
Place all certificates in the following store: click browser and select **Trusted Root Certificate Authorities**

Click Finish. **Restart Docker**.

![image](https://user-images.githubusercontent.com/2279508/193421112-4559cea9-dd96-4e72-ae58-7c0906819702.png)


ref here: https://docs.docker.com/registry/insecure/

**Comment by mduchain** [comment]
> maybe you no need to find the location on windows, but can install cert on windows by this way:
> 
> Windows Server:
> 
> Open Windows Explorer, right-click the ***.crt** file, and choose Install certificate. When prompted, select the following options:
> 
> Store location: local machine Place all certificates in the following store: click browser and select **Trusted Root Certificate Authorities**
> 
> Click Finish. **Restart Docker**.
> 
> ![image](https://user-images.githubusercontent.com/2279508/193421112-4559cea9-dd96-4e72-ae58-7c0906819702.png)
> 
> ref here: https://docs.docker.com/registry/insecure/

Just want to comment that the solution provided by @minhhungit works perfectly.
My situation: Windows  Server 2022 with docker installed following procedure in
https://learn.microsoft.com/en-us/virtualization/windowscontainers/quick-start/set-up-environment?tabs=dockerce#windows-server-1
Note: This is after the May 23 shutdown of [MicrosoftDockerProvider] service
(see https://github.com/OneGet/MicrosoftDockerProvider)

Important: restart the docker service!!
